### DISCUSSION TITLE
How to apply Gray-Scale filter?

### DISCUSSION THREAD

**Discussion by H1manshu21** [Blaze-Stars/Image-Filter-Application#25] [Q&A] [answered]
Discuss algorithm on how to apply `gray-scale` effect on `BMP` image.

**Comment by chandrakant100** [comment]
| Original Image V/S Grayscale image |
| ---------------------------------------- |
| <img src="http://ginva.com/wp-content/uploads/2018/01/img_5a69d9ae380e7.png" height="300px" width="500px"> |

**Reply by SHembram18** [reply]
**Algorithm for grayscale**
https://www.johndcook.com/blog/2009/08/24/algorithms-convert-color-grayscale/

**Reply by H1manshu21** [reply]
This is a good start to know how to show the intensity or scale of gray-scale.

**Reply by Ank1taS** [reply]
According to the article given by @HembramBeta777 [this](https://www.johndcook.com/blog/2009/08/24/algorithms-convert-color-grayscale/)
there are 3 different algorithms for ```Grayscale``` effects.

### Applying 3 different algorithms over an image
#### Original image
![original](https://user-images.githubusercontent.com/65461370/169100158-17640fce-96ad-4b9f-8a79-bc1aab69d1ac.png)

#### Grayscale - ```lightness algorithm``` 
Gives a lighter effect

#### Grayscale - ```average algorithm```
![avg](https://user-images.githubusercontent.com/65461370/169100569-38849ec6-5d21-4da2-bf97-b4cf5783b015.png)

#### Grayscale - ```luminosity algorithm```
![yard](https://user-images.githubusercontent.com/65461370/169100699-d2daf093-e2fe-4221-a121-2ba6395f8f9f.png)
Somehow luminosity algorithm doesn't work properly

**Reply by H1manshu21** [reply]
Consider this answer in another section and not in the reply section.

**Comment by Ank1taS** [comment, marked as answer]
By applying the ```Grayscale``` effect the original image will convert to a black and white image.

In GrayScale effect **all channels (red, green, blue) of a pixel have a specific value that, results in gray and according to this value, various shades of gray a pixel can have. Higher values mean lighter shades (closer to white) and lower values mean darker shades (closer to black)**.

There are different algorithms for Grayscale, however here is an illustration of ```Average algorithm```
In the **average algorithm, in a Grayscale image, each pixel's red, green and blue channel contains the same value i.e the average of the 3 channels of the original image**.

### Example of Original Image vs Grayscale Image
![gray](https://user-images.githubusercontent.com/65461370/169117920-67e2d114-2fa5-4d9e-a341-a9708fa47b15.jpg)

### Pseudocode for Grayscale filter

```
function reflection (Argumen1t rows, Argument2 columns, Argument3 pixel 2Darray ) {
        iterate and pick each row of the pixel 2D array
                iterate and pick each pixel of the row picked above
                        calculate and store the average of 3 channels (red, green, blue) of the picked pixel. 
                        store the calculated average value in every channel of the pixel.  
end
}
```

**Reply by H1manshu21** [reply]
Accepted to go with the average grayscale rather than the rest of the 2 gray-scale options. What do you think? @chandrakant100 @sabyasachi07 @Tushar7-coder @HembramBeta777

**Reply by SHembram18** [reply]
yes the average grayscale looks better than other two.
 @Himanshu40
